File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int solve(vector<int>& coins, int amount, vector<int>& dp) {
        if (amount == 0) {
            return 0;
        }
        
        if (dp[amount] != -1) {
            return dp[amount];
        }

        int minCoins = INT_MAX;

        for (int i = 0; i < coins.size(); i++) {
            if (amount - coins[i] >= 0 && dp[amount-coins[i]]!=INT_MAX) {
                int subproblem = solve(coins, amount - coins[i], dp);
                if (subproblem != -1) {
                    minCoins = min(minCoins, 1 + subproblem);
                }
            }
        }

        dp[amount] = (minCoins == INT_MAX) ? -1 : minCoins;
        return dp[amount];
    }

    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount + 1, -1);
        // return solve(coins, amount, dp);
        dp[0]=0;
        for(int i=1;i<=amount;i++){
            dp[i]=INT_MAX;
            for(int j=0;j<coins.size();j++){
                if(i-coins[j]>=0 && dp[i-coins[j]]!=INT_MAX){
                    dp[i]=min(dp[i],dp[i-coins[j]]+1);
                }
            }
        }
        return dp[amount]==INT_MAX?-1:dp[amount];
    }
};

File: 0322-coin-change/0322-coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int[][] memo = new int[coins.length][amount + 1];
        for (int i = 0; i < coins.length; i++) {
            for (int j = 0; j <= amount; j++) {
                memo[i][j] = -1;
            }
        }
        int result = solve(coins, amount, coins.length - 1, memo);
        return result == Integer.MAX_VALUE ? -1 : result;
    }

    public int solve(int[] coins, int amount, int i, int[][] memo) {
        if (amount == 0) {
            return 0;
        }
        if (i < 0) {
            return Integer.MAX_VALUE;
        }
        if (memo[i][amount] != -1) {
            return memo[i][amount];
        }
        int nonPick = solve(coins, amount, i - 1, memo);
        int pick = Integer.MAX_VALUE;
        if (amount >= coins[i]) {
            int res = solve(coins, amount - coins[i], i, memo);
            if (res != Integer.MAX_VALUE) {
                pick = 1 + res;
            }
        }
        memo[i][amount] = Math.min(pick, nonPick);
        return memo[i][amount];
    }
}
